include/deal.II/fe/fe_values.h

```diff
@@ -2911,6 +2911,84 @@ class FEValuesBase : public Subscriptor
+   *
+   *   for (auto &cell : dof_handler.active_cell_iterators())
+   *     {
+   *       fe_values.reinit(cell);
+   *       for (const auto i : fe_values.local_dof_indices())
+   *         for (const auto j : fe_values.local_dof_indices())
+   *           ... do something for DoF indices (i,j) ...
+   *     }
+   * @endcode
+   * Here, we are looping over all degrees of freedom on all cells, with
+   * `i` and `j` taking on all valid indices for cell degrees of freedom, as
+   * defined by the finite element passed to `fe_values`.
+   */
+  boost::integer_range<unsigned int>
+  local_dof_indices() const;
+
+  /**
+   * Return an object that can be thought of as an array containing all
+   * indices from @p start_dof_index to `dofs_per_cell`. This allows to write
+   * code using range-based for loops of the following kind:
+   * @code
+   *   FEValues<dim> fe_values (...);
+   *
+   *   for (auto &cell : dof_handler.active_cell_iterators())
+   *     {
+   *       fe_values.reinit(cell);
+   *       for (const auto i : fe_values.local_dof_indices())
+   *         for (const auto j : fe_values.local_dof_indices_starting_at(i))
+   *           ... do something for DoF indices (i,j) ...
+   *     }
+   * @endcode
+   * Here, we are looping over all local degrees of freedom on all cells, with
+   * `i` taking on all valid indices for cell degrees of freedom, as
+   * defined by the finite element passed to `fe_values`, and `j` taking
+   * on a specified subset of `i`'s range, starting at `i` itself and ending at
+   * the number of cell degrees of freedom. In this way, we can construct the
+   * upper half and the diagonal of a stiffness matrix contribution, for
+   * example.
+   */
+  boost::integer_range<unsigned int>
+  local_dof_indices_starting_at(const unsigned int start_dof_index) const;
+
+  /**
+   * Return an object that can be thought of as an array containing all
+   * indices from zero to @p end_dof_index. This allows to write code
+   * using range-based for loops of the following kind:
+   * @code
+   *   FEValues<dim> fe_values (...);
+   *
+   *   for (auto &cell : dof_handler.active_cell_iterators())
+   *     {
+   *       fe_values.reinit(cell);
+   *       for (const auto i : fe_values.local_dof_indices())
+   *         for (const auto j : fe_values.local_dof_indices_ending_at(i))
+   *           ... do something for DoF indices (i,j) ...
+   *     }
+   * @endcode
+   * Here, we are looping over all local degrees of freedom on all cells, with
+   * `i` taking on all valid indices for cell degrees of freedom, as
+   * defined by the finite element passed to `fe_values`, and `j` taking
+   * on a specified subset of `i`'s range, starting at zero and ending at
+   * `i` itself. In this way, we can construct the lower half and the
+   * diagonal of a stiffness matrix contribution, for example.
+   */
+  boost::integer_range<unsigned int>
+  local_dof_indices_ending_at(const unsigned int end_dof_index) const;
```

Comment:
I'm not sure how to answer this, since it is at the very least convenience function that someone may find useful to use. It hardly matters, but naturally it also decreases the number of loop iterations, as you can make the inner or outer loop counter _a priori_ smaller.

Here's examples of it in action. I like the convenience that it offers for the cases when one is assembling a symmetric matrix. I plan on using it, but if you feel strongly against it then we can drop it.
https://github.com/dealii/dealii/pull/9990/commits/451313af526c7ed1f4ee6b43f666c5d7fe702a52#diff-8a1c186228774bbd121920057a5b772cR1518-R1519
https://github.com/dealii/dealii/pull/9990/commits/451313af526c7ed1f4ee6b43f666c5d7fe702a52#diff-8a1c186228774bbd121920057a5b772cR1551-R1553